

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Example Jupyter Pipleine &mdash; PPP 0.1.7 documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="../../_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../" src="../../_static/documentation_options.js"></script>
        <script type="text/javascript" src="../../_static/jquery.js"></script>
        <script type="text/javascript" src="../../_static/underscore.js"></script>
        <script type="text/javascript" src="../../_static/doctools.js"></script>
        <script type="text/javascript" src="../../_static/language_data.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    
    <script type="text/javascript" src="../../_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="PPP Core Functions" href="../functions.html" />
    <link rel="prev" title="Examples" href="../examples.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../../index.html" class="icon icon-home"> PPP
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../intro.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="../install.html">Installation</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../examples.html">Examples</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../examples.html#function-examples">Function Examples</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="../examples.html#jupyter-notebook-pipelines">Jupyter Notebook Pipelines</a><ul class="current">
<li class="toctree-l3 current"><a class="current reference internal" href="#">Example Jupyter Pipleine</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#Setting-Filepaths">Setting Filepaths</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../functions.html">PPP Core Functions</a></li>
<li class="toctree-l1"><a class="reference internal" href="../input_file_generators.html">PPP Input File Generators</a></li>
<li class="toctree-l1"><a class="reference internal" href="../analyses.html">PPP Analyses</a></li>
<li class="toctree-l1"><a class="reference internal" href="../utilities.html">PPP Utilities</a></li>
<li class="toctree-l1"><a class="reference internal" href="../model.html">Model File and Creation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../develop.html">Development</a></li>
<li class="toctree-l1"><a class="reference internal" href="../contact.html">Contact Us</a></li>
<li class="toctree-l1"><a class="reference internal" href="../citations.html">Citations</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">PPP</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../index.html">Docs</a> &raquo;</li>
        
          <li><a href="../examples.html">Examples</a> &raquo;</li>
        
      <li>Example Jupyter Pipleine</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../../_sources/PPP_pages/jupyter/example_pipeline_pan.ipynb.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  
<style>
/* CSS for nbsphinx extension */

/* remove conflicting styling from Sphinx themes */
div.nbinput.container,
div.nbinput.container div.prompt,
div.nbinput.container div.input_area,
div.nbinput.container div[class*=highlight],
div.nbinput.container div[class*=highlight] pre,
div.nboutput.container,
div.nboutput.container div.prompt,
div.nboutput.container div.output_area,
div.nboutput.container div[class*=highlight],
div.nboutput.container div[class*=highlight] pre {
    background: none;
    border: none;
    padding: 0 0;
    margin: 0;
    box-shadow: none;
}

/* avoid gaps between output lines */
div.nboutput.container div[class*=highlight] pre {
    line-height: normal;
}

/* input/output containers */
div.nbinput.container,
div.nboutput.container {
    display: -webkit-flex;
    display: flex;
    align-items: flex-start;
    margin: 0;
    width: 100%;
}
@media (max-width: 540px) {
    div.nbinput.container,
    div.nboutput.container {
        flex-direction: column;
    }
}

/* input container */
div.nbinput.container {
    padding-top: 5px;
}

/* last container */
div.nblast.container {
    padding-bottom: 5px;
}

/* input prompt */
div.nbinput.container div.prompt pre {
    color: #307FC1;
}

/* output prompt */
div.nboutput.container div.prompt pre {
    color: #BF5B3D;
}

/* all prompts */
div.nbinput.container div.prompt,
div.nboutput.container div.prompt {
    width: 4.5ex;
    padding-top: 5px;
    position: relative;
    user-select: none;
}

div.nbinput.container div.prompt > div,
div.nboutput.container div.prompt > div {
    position: absolute;
    right: 0;
    margin-right: 0.3ex;
}

@media (max-width: 540px) {
    div.nbinput.container div.prompt,
    div.nboutput.container div.prompt {
        width: unset;
        text-align: left;
        padding: 0.4em;
    }
    div.nboutput.container div.prompt.empty {
        padding: 0;
    }

    div.nbinput.container div.prompt > div,
    div.nboutput.container div.prompt > div {
        position: unset;
    }
}

/* disable scrollbars on prompts */
div.nbinput.container div.prompt pre,
div.nboutput.container div.prompt pre {
    overflow: hidden;
}

/* input/output area */
div.nbinput.container div.input_area,
div.nboutput.container div.output_area {
    -webkit-flex: 1;
    flex: 1;
    overflow: auto;
}
@media (max-width: 540px) {
    div.nbinput.container div.input_area,
    div.nboutput.container div.output_area {
        width: 100%;
    }
}

/* input area */
div.nbinput.container div.input_area {
    border: 1px solid #e0e0e0;
    border-radius: 2px;
    background: #f5f5f5;
}

/* override MathJax center alignment in output cells */
div.nboutput.container div[class*=MathJax] {
    text-align: left !important;
}

/* override sphinx.ext.imgmath center alignment in output cells */
div.nboutput.container div.math p {
    text-align: left;
}

/* standard error */
div.nboutput.container div.output_area.stderr {
    background: #fdd;
}

/* ANSI colors */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-default-inverse-fg { color: #FFFFFF; }
.ansi-default-inverse-bg { background-color: #000000; }

.ansi-bold { font-weight: bold; }
.ansi-underline { text-decoration: underline; }


div.nbinput.container div.input_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight].math,
div.nboutput.container div.output_area.rendered_html,
div.nboutput.container div.output_area > div.output_javascript,
div.nboutput.container div.output_area:not(.rendered_html) > img{
    padding: 5px;
}

/* fix copybtn overflow problem in chromium (needed for 'sphinx_copybutton') */
div.nbinput.container div.input_area > div[class^='highlight'],
div.nboutput.container div.output_area > div[class^='highlight']{
    overflow-y: hidden;
}

/* hide copybtn icon on prompts (needed for 'sphinx_copybutton') */
.prompt a.copybtn {
    display: none;
}

/* Some additional styling taken form the Jupyter notebook CSS */
div.rendered_html table {
  border: none;
  border-collapse: collapse;
  border-spacing: 0;
  color: black;
  font-size: 12px;
  table-layout: fixed;
}
div.rendered_html thead {
  border-bottom: 1px solid black;
  vertical-align: bottom;
}
div.rendered_html tr,
div.rendered_html th,
div.rendered_html td {
  text-align: right;
  vertical-align: middle;
  padding: 0.5em 0.5em;
  line-height: normal;
  white-space: normal;
  max-width: none;
  border: none;
}
div.rendered_html th {
  font-weight: bold;
}
div.rendered_html tbody tr:nth-child(odd) {
  background: #f5f5f5;
}
div.rendered_html tbody tr:hover {
  background: rgba(66, 165, 245, 0.2);
}

/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<div class="section" id="Example-Jupyter-Pipleine">
<h1>Example Jupyter Pipleine<a class="headerlink" href="#Example-Jupyter-Pipleine" title="Permalink to this headline">¶</a></h1>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">sys</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">subprocess</span>

<span class="kn">from</span> <span class="nn">pgpipe</span> <span class="k">import</span> <span class="n">four_gamete</span><span class="p">,</span> <span class="n">vcf_split_pysam</span><span class="p">,</span> <span class="n">vcf_to_ima</span><span class="p">,</span> <span class="n">vcf_filter</span><span class="p">,</span> <span class="n">vcf_calc</span><span class="p">,</span> <span class="n">vcf_sampler</span><span class="p">,</span> <span class="n">vcf_phase</span><span class="p">,</span> <span class="n">stat_sampler</span><span class="p">,</span> <span class="n">vcf_split</span>
<span class="kn">from</span> <span class="nn">pgpipe.logging_module</span> <span class="k">import</span> <span class="n">initLogger</span>
<span class="kn">from</span> <span class="nn">pgpipe.informative_loci_filter</span> <span class="k">import</span> <span class="n">filter_bed_regions</span>
<span class="kn">from</span> <span class="nn">pgpipe.subtract_bed</span> <span class="k">import</span> <span class="n">filter_stat</span>
<span class="kn">import</span> <span class="nn">pysam</span>

<span class="nb">print</span> <span class="p">(</span><span class="s2">&quot;Imports complete&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Imports complete
</pre></div></div>
</div>
<div class="section" id="Setting-Filepaths">
<h2>Setting Filepaths<a class="headerlink" href="#Setting-Filepaths" title="Permalink to this headline">¶</a></h2>
<p>The required input files for a PPP run are: - A genome VCF of the target populations (plus a tabix index if bgzipped) - A population model file</p>
<p>The population model file is a JSON-formatted file that defines population names and the individuals from the VCF that belong to each population. This file can be created using the model_creator function, or by creating it manually by using an example model file as a template.</p>
<p>For region filtering, the following should be provided: - Name for target region file - Name for final selected region file - File with genic regions/regions to be excluded from analysis (optional) - File with regions to be selected from for analysis (optional)</p>
<p>There are two methods for obtaining target regions for subsampling and analysis, with some level of interoperability: using a statistic file to sample regions, or use a file with target regions to randomly select regions with enough sites to be valid in an IM analysis. To generate a statistic file, use vcf_calc to read over the input VCF file, then use stat_sampler to select either a random or uniform distribution of these regions given their statistic value. These stat files can be filtered
with a genic region file, available from UCSC, using the subtract_bed function. This method is implemented in this notebook, with the genic region filtering offered as an optional cell.</p>
<p>An additional method of region selection, done without statistics, can be done by downloading region files for genic regions, and optionally for STR regions and regions with missing data. If one wants to find a set of target regions that are intergenic and outside of STR regions, download the corresponding files from UCSC for your species and use invert_bed_region to ‘invert’ the files, optionally selecting only regions outside of a set number of bases from the regions in the file with the
–window option. If files don’t have typical BED column order, use the –bed-column-index option to provide a comma-separated string with the 0-based index of the start, end, and chromosome column. (For normal BED files, this would be ‘1,2,0’) The get_nonmissing_chunks function can scan the input VCF and find regions with no missing data. All of these can be combined with bedtools to select regions that overlap with all three possible files.</p>
<p>Whichever method used, the target region file should be run through the informative_filter function to check the VCF file has enough biallelic, informative SNPs (two or more of each of two alleles) to have a good chance of passing the four-gamete test (which requires regions with at least two SNPs). An informative count of 5 will usually allow for this, but if you have limited regions this threshhold can be lowered to 3.</p>
<p>In addition to these files, additional functionality such as CpG filtering and comprehensive logging can be used by providing: - A reference FASTA (for CpG filtering, can be bgzipped or unzipped but requires indexing w/faidx) - A log filename</p>
<p>This section also is used to set up the directory structure for data files, a working directory, a directory for VCF region files, a target number of loci, and names for various stages of loci VCF files (phasing, four-gamete testing, potentially filtering individuals with missing data).</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1">#Set up directories and filepaths, run on all restarts</span>
<span class="n">work_dir</span><span class="o">=</span><span class="s1">&#39;/home/jared/workspace/projects_ppp/notebook_sample/&#39;</span>
<span class="c1">#data_dir=&#39;/media/ccgg/ppp_sample_data/&#39;</span>
<span class="n">data_dir</span><span class="o">=</span><span class="n">work_dir</span>
<span class="n">vcf_dir</span> <span class="o">=</span> <span class="n">work_dir</span><span class="o">+</span><span class="s1">&#39;vcfs/&#39;</span>


<span class="n">main_vcf_name</span> <span class="o">=</span> <span class="n">data_dir</span><span class="o">+</span><span class="s1">&#39;pan_chr20.vcf.gz&#39;</span>
<span class="n">filtered_vcf_name</span> <span class="o">=</span> <span class="n">data_dir</span><span class="o">+</span><span class="s1">&#39;pan_chr20_filtered.vcf.gz&#39;</span>
<span class="n">stat_file_name</span> <span class="o">=</span> <span class="n">work_dir</span><span class="o">+</span><span class="s1">&#39;fst_regions.bed&#39;</span>
<span class="n">model_file</span> <span class="o">=</span> <span class="n">data_dir</span><span class="o">+</span><span class="s1">&#39;great_ape.model&#39;</span>
<span class="n">int_bed_file</span> <span class="o">=</span> <span class="n">work_dir</span><span class="o">+</span><span class="s1">&#39;regions_for_sampling.bed&#39;</span>
<span class="n">target_loci_file</span> <span class="o">=</span> <span class="n">work_dir</span><span class="o">+</span><span class="s1">&#39;target_loci.bed&#39;</span>
<span class="n">ima_input_file</span> <span class="o">=</span> <span class="n">work_dir</span><span class="o">+</span><span class="s1">&#39;test_run_input.ima.u&#39;</span>
<span class="c1">#subsamp_bed_file = work_dir+&#39;great_ape_genome2/5k_sample.bed&#39;</span>
<span class="n">logfile</span> <span class="o">=</span> <span class="s1">&#39;/home/jared/testpppj.log&#39;</span>

<span class="n">loci</span><span class="o">=</span><span class="mi">200</span>

<span class="n">region_files</span> <span class="o">=</span> <span class="p">[</span><span class="n">vcf_dir</span><span class="o">+</span><span class="s1">&#39;Sampled_nonmissing/Sample_&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span><span class="o">+</span><span class="s1">&#39;.vcf&#39;</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">loci</span><span class="p">)]</span>
<span class="n">phased_files</span> <span class="o">=</span> <span class="p">[</span><span class="n">vcf_dir</span><span class="o">+</span><span class="s1">&#39;Phased/phased_&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span><span class="o">+</span><span class="s1">&#39;.vcf&#39;</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">loci</span><span class="p">)]</span>
<span class="n">fourg_files</span> <span class="o">=</span> <span class="p">[</span><span class="n">vcf_dir</span><span class="o">+</span><span class="s1">&#39;four_gamete/Sample_&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span><span class="o">+</span><span class="s1">&#39;.vcf&#39;</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">loci</span><span class="p">)]</span>
<span class="n">passed_files</span> <span class="o">=</span> <span class="p">[]</span>

</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1">#Set up directory structure, only needs to be run once</span>
<span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">vcf_dir</span><span class="p">):</span>
    <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">vcf_dir</span><span class="p">)</span>
    <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">vcf_dir</span><span class="o">+</span><span class="s1">&#39;four_gamete/&#39;</span><span class="p">)</span>
    <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">vcf_dir</span><span class="o">+</span><span class="s1">&#39;Sampled_nonmissing/&#39;</span><span class="p">)</span>
    <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">vcf_dir</span><span class="o">+</span><span class="s1">&#39;Phased/&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>The vcf_filter step will filter the original VCF according to many conditions, including: - Non-biallelic sites (–filter-min-alleles and –filter-max-alleles) - Sites with missing data (–filter-max-missing) - Indels - Sites on non-autosomal chromosomes (–filter-exclude-chr) - Individuals not named in model file (use –model-file to input model file, with –model if multiple models included in file)</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1">#Creates VCF filtered for no missing data and biallelic sites</span>
<span class="n">vcf_filter</span><span class="o">.</span><span class="n">run</span><span class="p">([</span><span class="s1">&#39;--vcf&#39;</span><span class="p">,</span> <span class="n">main_vcf_name</span><span class="p">,</span> <span class="s1">&#39;--filter-max-missing&#39;</span><span class="p">,</span> <span class="s1">&#39;1.0&#39;</span><span class="p">,</span> <span class="s1">&#39;--model-file&#39;</span><span class="p">,</span><span class="n">model_file</span><span class="p">,</span>
                <span class="s1">&#39;--model&#39;</span><span class="p">,</span><span class="s1">&#39;2Pop&#39;</span><span class="p">,</span> <span class="s1">&#39;--filter-min-alleles&#39;</span><span class="p">,</span> <span class="s1">&#39;2&#39;</span><span class="p">,</span> <span class="s1">&#39;--filter-max-alleles&#39;</span><span class="p">,</span> <span class="s1">&#39;2&#39;</span><span class="p">,</span> <span class="s1">&#39;--out-format&#39;</span><span class="p">,</span>
                <span class="s1">&#39;vcf.gz&#39;</span><span class="p">,</span> <span class="s1">&#39;--out&#39;</span><span class="p">,</span> <span class="n">filtered_vcf_name</span><span class="p">,</span> <span class="s1">&#39;--filter-exclude-chr&#39;</span><span class="p">,</span> <span class="s1">&#39;chrX&#39;</span><span class="p">,</span> <span class="s1">&#39;chrY&#39;</span><span class="p">,</span> <span class="s1">&#39;--overwrite&#39;</span><span class="p">])</span>

<span class="n">pysam</span><span class="o">.</span><span class="n">tabix_index</span><span class="p">(</span><span class="n">filtered_vcf_name</span><span class="p">,</span><span class="n">preset</span><span class="o">=</span><span class="s1">&#39;vcf&#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Filtering complete&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Filtering complete
</pre></div></div>
</div>
<p>The vcf_calc step will, for every 10kb window in the genome, calculate Fst given populations from the model file. Statistics that can be filtered over currently include: - Windowed pi - Tajima’s D - Fst</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1">#Calculates f_st statistics across genome</span>
<span class="n">vcf_calc</span><span class="o">.</span><span class="n">run</span><span class="p">([</span><span class="s1">&#39;--vcf&#39;</span><span class="p">,</span> <span class="n">filtered_vcf_name</span><span class="p">,</span> <span class="s1">&#39;--out&#39;</span><span class="p">,</span> <span class="n">stat_file_name</span><span class="p">,</span>
              <span class="s1">&#39;--calc-statistic&#39;</span><span class="p">,</span> <span class="s1">&#39;windowed-weir-fst&#39;</span><span class="p">,</span> <span class="s1">&#39;--model&#39;</span><span class="p">,</span> <span class="s1">&#39;2Pop&#39;</span><span class="p">,</span> <span class="s1">&#39;--statistic-window-size&#39;</span><span class="p">,</span>
              <span class="s1">&#39;10000&#39;</span><span class="p">,</span> <span class="s1">&#39;--statistic-window-step&#39;</span><span class="p">,</span> <span class="s1">&#39;10000&#39;</span><span class="p">,</span> <span class="s1">&#39;--model-file&#39;</span><span class="p">,</span> <span class="n">model_file</span><span class="p">,</span> <span class="s1">&#39;--overwrite&#39;</span><span class="p">])</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Stat calculation complete&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Stat calculation complete
</pre></div></div>
</div>
<p>Using the informative site filter, the regions produced by the statistics generation can be checked for whether they contain enough informative sites to pass the four-gamete test. Additional filtering can be done here if it hasn’t been done before.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1">#Selects subset of regions for fast sampling</span>
<span class="n">filter_bed_regions</span><span class="p">([</span><span class="s1">&#39;--vcf&#39;</span><span class="p">,</span><span class="n">filtered_vcf_name</span><span class="p">,</span><span class="s1">&#39;--bed&#39;</span><span class="p">,</span><span class="n">stat_file_name</span><span class="p">,</span>
                    <span class="s1">&#39;--remove-indels&#39;</span><span class="p">,</span><span class="s1">&#39;--minsites&#39;</span><span class="p">,</span><span class="s1">&#39;3&#39;</span><span class="p">,</span><span class="s1">&#39;--keep-full-line&#39;</span><span class="p">,</span><span class="s1">&#39;--out&#39;</span><span class="p">,</span><span class="n">int_bed_file</span><span class="p">,</span>
                    <span class="s1">&#39;--randcount&#39;</span><span class="p">,</span><span class="s1">&#39;5000&#39;</span><span class="p">,</span><span class="s1">&#39;--remove-multi&#39;</span><span class="p">])</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;BED regions selected&quot;</span><span class="p">)</span>

</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
BED regions selected
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Only 2848 of 5000 regions found
</pre></div></div>
</div>
<p>(Optional) If a file with genic regions is provided, statistic windows that overlap those regions can be removed from potential loci. The window option can be used to extend exclusion regions by a set number of base pairs up AND downstream of regions in the filter file. Zero-ho indicates that files use a zero-based, half open interval representation, as opposed to the general 1-based, closed region format.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">int_bed_file2</span> <span class="o">=</span> <span class="n">work_dir</span><span class="o">+</span><span class="s1">&#39;regions_for_sampling_nogenes.bed&#39;</span>
<span class="n">gene_file</span> <span class="o">=</span> <span class="n">work_dir</span><span class="o">+</span><span class="s1">&#39;hg18_chr22_genes.bed&#39;</span>
<span class="n">filter_stat</span><span class="p">([</span><span class="s1">&#39;--stat-file&#39;</span><span class="p">,</span><span class="n">int_bed_file</span><span class="p">,</span><span class="s1">&#39;--filter-file&#39;</span><span class="p">,</span><span class="n">gene_file</span><span class="p">,</span><span class="s1">&#39;--window&#39;</span><span class="p">,</span><span class="s1">&#39;10000&#39;</span><span class="p">,</span><span class="s1">&#39;--zero-ho&#39;</span><span class="p">,</span>
            <span class="s1">&#39;--out&#39;</span><span class="p">,</span><span class="n">int_bed_file2</span><span class="p">])</span>
<span class="n">int_bed_file</span> <span class="o">=</span> <span class="n">int_bed_file2</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
WARNING:root:1773 of 2848 regions selected as non-overlapping
</pre></div></div>
</div>
<p>The statistic file can be filtered in one of two ways: randomly or uniformly. A random sample (which conceptually doesn’t require a statistic) will select sample-size numer of loci for analysis, while a uniform sample will attempt to create a uniform distribution of the chosen statistic. The samples will be placed into a set number of bins, which must be divisible by the number of target loci.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">stat_sampler</span><span class="o">.</span><span class="n">run</span><span class="p">([</span><span class="s1">&#39;--statistic-file&#39;</span><span class="p">,</span><span class="n">int_bed_file</span><span class="p">,</span><span class="s1">&#39;--out&#39;</span><span class="p">,</span><span class="n">target_loci_file</span><span class="p">,</span><span class="s1">&#39;--sampling-scheme&#39;</span><span class="p">,</span><span class="s1">&#39;uniform&#39;</span><span class="p">,</span>
                 <span class="s1">&#39;--uniform-bins&#39;</span><span class="p">,</span><span class="s1">&#39;5&#39;</span><span class="p">,</span><span class="s1">&#39;--sample-size&#39;</span><span class="p">,</span><span class="nb">str</span><span class="p">(</span><span class="n">loci</span><span class="p">),</span><span class="s1">&#39;--calc-statistic&#39;</span><span class="p">,</span><span class="s1">&#39;windowed-weir-fst&#39;</span><span class="p">,</span><span class="s1">&#39;--overwrite&#39;</span><span class="p">])</span>
</pre></div>
</div>
</div>
<p>This function creates loci VCF files from the full-genome VCF, while optionally doing additional filtering</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1">#Uniformly sample regions for subset of 200 loci</span>
<span class="c1">#vcf_sampler.run([&#39;--vcf&#39;, filtered_vcf_name, &#39;--statistic-file&#39;,</span>
<span class="c1">#                 target_loci_file, &#39;--out-format&#39;, &#39;vcf&#39;, &#39;--calc-statistic&#39;, &#39;windowed-weir-fst&#39;,</span>
<span class="c1">#                 &#39;--sampling-scheme&#39;, &#39;uniform&#39;, &#39;--uniform-bins&#39;, &#39;5&#39;, &#39;--out-dir&#39;,</span>
<span class="c1">#                 work_dir + &#39;great_ape_genome2/Sample_Files&#39;, &#39;--overwrite&#39;])</span>
<span class="c1">#vcf_split.run([&#39;--vcf&#39;,filtered_vcf_name,&#39;--split-method&#39;,&#39;statistic-file&#39;,&#39;--out-format&#39;,&#39;vcf&#39;,&#39;--out-prefix&#39;,</span>
<span class="c1">#              vcf_dir+&#39;Sampled_nonmissing/Sample&#39;,&#39;--split-file&#39;,target_loci_file])</span>
<span class="n">vcf_split_pysam</span><span class="o">.</span><span class="n">vcf_region_write</span><span class="p">([</span><span class="n">filtered_vcf_name</span><span class="p">,</span><span class="s1">&#39;--bed&#39;</span><span class="p">,</span><span class="n">target_loci_file</span><span class="p">,</span><span class="s1">&#39;--out-prefix&#39;</span><span class="p">,</span><span class="n">vcf_dir</span><span class="o">+</span><span class="s1">&#39;Sampled_nonmissing/Sample_&#39;</span><span class="p">,</span>
                                  <span class="s1">&#39;--remove-indels&#39;</span><span class="p">,</span><span class="s1">&#39;--remove-multi&#39;</span><span class="p">,</span><span class="s1">&#39;--bed-column-index&#39;</span><span class="p">,</span><span class="s1">&#39;2,3,1&#39;</span><span class="p">,</span>
                                   <span class="s1">&#39;--informative-count&#39;</span><span class="p">,</span><span class="s1">&#39;2&#39;</span><span class="p">])</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Sampling complete&quot;</span><span class="p">)</span>

</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Sampling complete
</pre></div></div>
</div>
<p>Each locus must be prepared for IM analysis, which involves finding a subregion of each locus that passes the four-gamete test. The four-gamete test is passed if all pairs of alleles in a region have less than four gametes among them. For example, if two SNPs are A/C and G/T, there are four possible gamete haplotypes among them: AG, AT, CG, and CT. If haplotypes with all four of these are present, this indicates that there must have been a recombination event between them at some point in the
sampled populations. This violates the IM model, so these regions would fail the test. The four-gamete code as implemented will compute all regions that pass the four-gamete test in a locus VCF file, then select a region either at random or with the largest number of informative sites. A minimum number of informative sites can be set, which defaults to two.</p>
<p>Before the four-gamete test, the locus VCF files need to be phased. This pipeline provides two phasing programs, beagle and shapeit.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1">#Phase locus</span>
<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">loci</span><span class="p">):</span>
    <span class="n">vcf_phase</span><span class="o">.</span><span class="n">run</span><span class="p">([</span><span class="s1">&#39;--vcf&#39;</span><span class="p">,</span><span class="n">region_files</span><span class="p">[</span><span class="n">i</span><span class="p">],</span><span class="s1">&#39;--phase-algorithm&#39;</span><span class="p">,</span><span class="s1">&#39;shapeit&#39;</span><span class="p">,</span><span class="s1">&#39;--out&#39;</span><span class="p">,</span>
                   <span class="n">phased_files</span><span class="p">[</span><span class="n">i</span><span class="p">],</span><span class="s1">&#39;--out-format&#39;</span><span class="p">,</span><span class="s1">&#39;vcf&#39;</span><span class="p">,</span><span class="s1">&#39;--overwrite&#39;</span><span class="p">])</span>
<span class="nb">print</span> <span class="p">(</span><span class="s2">&quot;Phasing done&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Phasing done
</pre></div></div>
</div>
<p>Once phasing is done, each file must be filtered through the four-gamete test. The four-gamete test is a method for determining whether or not there has been recombination between a pair of variants. To do this, all individuals have haplotypes defined as the variants at the two sites. Given two snps with ref/alt alleles A/G and C/T, if individuals in this sample have haplotypes AC, AT, and GT, it is possible that there has been no recombination between these alleles. If an additional individual
has the GC haplotype, this means that a recombination event must have taken place between the sites. This function will return a subregion of the region in the contained VCF that passes the four-gamete test with at least two informative (ac&gt;1) SNPs. If no valid region is found, no VCF is created and the region is skipped for downstream analysis.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1">#Subsample locus for four-gamete compatible interval, if no subregion returned, do not use VCF</span>
<span class="n">passed_files</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">loci</span><span class="p">):</span>
    <span class="n">ret</span> <span class="o">=</span> <span class="n">four_gamete</span><span class="o">.</span><span class="n">sample_fourgametetest_intervals</span><span class="p">([</span><span class="s1">&#39;--vcfs&#39;</span><span class="p">,</span> <span class="n">phased_files</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="s1">&#39;--out&#39;</span><span class="p">,</span>
                                                       <span class="n">fourg_files</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="s1">&#39;--4gcompat&#39;</span><span class="p">,</span> <span class="s1">&#39;--reti&#39;</span><span class="p">,</span> <span class="s1">&#39;--right&#39;</span><span class="p">,</span>
                                                       <span class="s1">&#39;--numinf&#39;</span><span class="p">,</span> <span class="s1">&#39;2&#39;</span><span class="p">])</span>
    <span class="k">if</span> <span class="n">ret</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">passed_files</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">fourg_files</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>
<span class="nb">print</span> <span class="p">(</span><span class="s2">&quot;Four gamete regions selected for </span><span class="si">%d</span><span class="s2"> loci&quot;</span><span class="o">%</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">passed_files</span><span class="p">)))</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Four gamete regions selected for 199 loci
</pre></div></div>
</div>
<p>This converts the files that pass the four-gamete test into a single IMa input file. Required arguments are a list of VCF files, and a model file. Filtering options are also available if unwanted sites haven’t been filtered out at a previous step.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1">#Create IMa input file</span>
<span class="n">ima_args</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;--vcfs&#39;</span><span class="p">]</span>
<span class="n">ima_args</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">passed_files</span><span class="p">)</span>
<span class="n">ima_args</span><span class="o">.</span><span class="n">extend</span><span class="p">([</span><span class="s1">&#39;--model-file&#39;</span><span class="p">,</span> <span class="n">model_file</span><span class="p">,</span> <span class="s1">&#39;--model&#39;</span><span class="p">,</span><span class="s1">&#39;2Pop&#39;</span><span class="p">,</span><span class="s1">&#39;--out&#39;</span><span class="p">,</span> <span class="n">work_dir</span> <span class="o">+</span> <span class="s1">&#39;ima_all_loci.ima.u&#39;</span><span class="p">])</span>

<span class="n">vcf_to_ima</span><span class="o">.</span><span class="n">vcf_to_ima</span><span class="p">(</span><span class="n">ima_args</span><span class="p">)</span>
<span class="nb">print</span> <span class="p">(</span><span class="s2">&quot;IMa input created&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
IMa input created
</pre></div></div>
</div>
<p>If desired, this block will run admixture to determine the population assignments of the various populations in the input VCF files. The plot will indicate, for each individual, an estimate of how much of their ancestry comes from the populations determined by clustering in admixture.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1">#Admixture analysis, optional</span>
<span class="kn">from</span> <span class="nn">pgpipe</span> <span class="k">import</span> <span class="n">convert</span><span class="p">,</span> <span class="n">admixture</span><span class="p">,</span> <span class="n">graph_plotter</span>
<span class="n">phased_string</span> <span class="o">=</span> <span class="s1">&#39; &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">phased_files</span><span class="p">)</span>
<span class="n">loci_vcf</span> <span class="o">=</span> <span class="n">vcf_dir</span><span class="o">+</span><span class="s1">&#39;Phased/phased_merged.vcf.gz&#39;</span>
<span class="n">concatcall</span> <span class="o">=</span> <span class="n">subprocess</span><span class="o">.</span><span class="n">Popen</span><span class="p">(</span><span class="s1">&#39;vcf-concat &#39;</span><span class="o">+</span><span class="n">phased_string</span><span class="o">+</span> <span class="s1">&#39; | bgzip -c &gt; &#39;</span><span class="o">+</span><span class="n">loci_vcf</span><span class="p">,</span> <span class="n">shell</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span><span class="n">stdout</span><span class="o">=</span><span class="n">subprocess</span><span class="o">.</span><span class="n">PIPE</span><span class="p">)</span>
<span class="n">temp_out</span><span class="p">,</span> <span class="n">temp_err</span> <span class="o">=</span> <span class="n">concatcall</span><span class="o">.</span><span class="n">communicate</span><span class="p">()</span>
<span class="n">convert</span><span class="o">.</span><span class="n">run</span><span class="p">([</span><span class="s1">&#39;--vcf&#39;</span><span class="p">,</span><span class="n">loci_vcf</span><span class="p">,</span><span class="s1">&#39;--out-format&#39;</span><span class="p">,</span><span class="s1">&#39;binary-ped&#39;</span><span class="p">,</span><span class="s1">&#39;--out-prefix&#39;</span><span class="p">,</span><span class="n">vcf_dir</span><span class="o">+</span><span class="s1">&#39;great_ape&#39;</span><span class="p">,</span><span class="s1">&#39;--overwrite&#39;</span><span class="p">])</span>
<span class="n">admixture</span><span class="o">.</span><span class="n">run</span><span class="p">([</span><span class="s1">&#39;--file&#39;</span><span class="p">,</span><span class="n">vcf_dir</span><span class="o">+</span><span class="s1">&#39;great_ape.bed&#39;</span><span class="p">,</span><span class="s1">&#39;--pop&#39;</span><span class="p">,</span><span class="s1">&#39;2&#39;</span><span class="p">])</span>
<span class="n">graph_plotter</span><span class="o">.</span><span class="n">bar_plot</span><span class="p">(</span><span class="n">vcf_dir</span><span class="o">+</span><span class="s1">&#39;great_ape.2.Q&#39;</span><span class="p">)</span>
<span class="nb">print</span> <span class="p">(</span><span class="s2">&quot;Plots created&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span>
</pre></div>
</div>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="../functions.html" class="btn btn-neutral float-right" title="PPP Core Functions" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="../examples.html" class="btn btn-neutral float-left" title="Examples" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2020, PPP Development Team. This material is based upon work supported by the National Science Foundation under Grant Number (NSF ABI 1564659 to Arun Sethuraman and Jody Hey)

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>