

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>vcf_to_sfs.py: Site Frequency Spectrum generator &mdash; PPP 0.1.7 documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="../../_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../" src="../../_static/documentation_options.js"></script>
        <script type="text/javascript" src="../../_static/jquery.js"></script>
        <script type="text/javascript" src="../../_static/underscore.js"></script>
        <script type="text/javascript" src="../../_static/doctools.js"></script>
        <script type="text/javascript" src="../../_static/language_data.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    
    <script type="text/javascript" src="../../_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="PPP Utilities" href="../utilities.html" />
    <link rel="prev" title="plink_linkage_disequilibrium.py: Linkage Disequilibrium Analysis" href="plink_linkage_disequilibrium.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../../index.html" class="icon icon-home"> PPP
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../intro.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="../install.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../examples.html">Examples</a></li>
<li class="toctree-l1"><a class="reference internal" href="../functions.html">PPP Core Functions</a></li>
<li class="toctree-l1"><a class="reference internal" href="../input_file_generators.html">PPP Input File Generators</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../analyses.html">PPP Analyses</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="eigenstrat_fstats.html">eigenstrat_fstats.py: F-statistics Analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="admixture.html">admixture.py: Admixture Analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="ima3_wrapper.html">ima3_wrapper.py: IMa3 Analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="plink_linkage_disequilibrium.html">plink_linkage_disequilibrium.py: Linkage Disequilibrium Analysis</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">vcf_to_sfs.py: Site Frequency Spectrum generator</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#required-arguments">Required Arguments</a></li>
<li class="toctree-l3"><a class="reference internal" href="#optional-aguments">Optional Aguments</a></li>
<li class="toctree-l3"><a class="reference internal" href="#example-usage">Example usage</a></li>
<li class="toctree-l3"><a class="reference internal" href="#importing-functions">Importing Functions</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#build-sfs">build_sfs()</a></li>
<li class="toctree-l4"><a class="reference internal" href="#reduce-sfs-dims">reduce_sfs_dims()</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../utilities.html">PPP Utilities</a></li>
<li class="toctree-l1"><a class="reference internal" href="../model.html">Model File and Creation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../develop.html">Development</a></li>
<li class="toctree-l1"><a class="reference internal" href="../contact.html">Contact Us</a></li>
<li class="toctree-l1"><a class="reference internal" href="../citations.html">Citations</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">PPP</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../index.html">Docs</a> &raquo;</li>
        
          <li><a href="../analyses.html">PPP Analyses</a> &raquo;</li>
        
      <li>vcf_to_sfs.py: Site Frequency Spectrum generator</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../../_sources/PPP_pages/Analyses/vcf_to_sfs.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  
<style>
/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<div class="section" id="module-vcf_to_sfs">
<span id="vcf-to-sfs-py-site-frequency-spectrum-generator"></span><h1>vcf_to_sfs.py: Site Frequency Spectrum generator<a class="headerlink" href="#module-vcf_to_sfs" title="Permalink to this headline">¶</a></h1>
<p>For generating the site frequency spectrum (sfs)for a
population model from a vcf file.</p>
<p>The sfs is an array with as many dimensions as populations in the model.
For example, if population samples are in order A,B, C
then position (i,j,k) of the array refers to the count of SNPs with derived alleles that
were observed to have a count of i in A, j in B, and k in C</p>
<p>If the sfs is folded then the count in a cell of the sfs is the number
of SNPs with that combination of minor allele counts.</p>
<p>This script can be run in stand-alone mode,  or for more flexibility it can
be imported to give access to more general functions for building and
manipulating the sfs.</p>
<p>All vcf handling assumes that all individuals are diploid at all SNPs.</p>
<div class="section" id="required-arguments">
<h2>Required Arguments<a class="headerlink" href="#required-arguments" title="Permalink to this headline">¶</a></h2>
<dl class="simple">
<dt><strong>--vcf</strong> <em>&lt;input_vcf_filename&gt;</em></dt><dd><p>The name of the vcf file.  This can be a bgzipped vcf file. .</p>
</dd>
<dt><strong>--model-file</strong> <em>&lt;model_file_name&gt;</em></dt><dd><p>The name of a PPP model file.</p>
</dd>
<dt><strong>--modelname</strong> <em>&lt;model_name&gt;</em></dt><dd><p>The name of a model in the model file.  The treemix file to be
generated will contain the allele counts for each SNP in each of
the populations.  The treemix run will estimate the phylogeny
for the populations in the model.</p>
</dd>
<dt><strong>--out</strong> <em>&lt;out file name&gt;</em></dt><dd><p>The name of an output file. If --out is omitted the default is ppp_sfs.out in
the same folder as the vcffile This will be a tab-delimited file
If the number of dimensions is 2, the sfs is contained in the rows and columns,
otherwise the values are given on the first line of the file</p>
</dd>
</dl>
</div>
<div class="section" id="optional-aguments">
<h2>Optional Aguments<a class="headerlink" href="#optional-aguments" title="Permalink to this headline">¶</a></h2>
<dl>
<dt><strong>--bed-file</strong> <em>&lt;BED_file_name&gt;</em></dt><dd><p>The BED file is a sorted UCSC-style bedfile containing chromosome locations of
the SNPs to be included in the output files. The BED file has no header.
The first column is the chromosome name (this must match the chromosome
name in the vcf file).
The second column is start position (0-based, open interval)
The third column is end position (closed interval).
Any other columns are ignored.</p>
</dd>
<dt><strong>--outgroup_fasta</strong> <em>&lt;name of alternative reference sequence&gt;</em></dt><dd><p>This option is used to specify the name of a fasta file to use as an
alternative reference to that originally used for the vcf file.</p>
<p>This fasta file must have been properly aligned to the reference
used in the vcf file.</p>
<p>This option can be useful, for example, if an ancestral or outgroup reference is
available that more accurately identifies the ancestral (and thus derived)
allele at each SNP than does the reference used to make the vcf file.</p>
</dd>
<dt><strong>--downsamplesizes</strong> <em>&lt;down sample sizes&gt;</em></dt><dd><p>A sequence of integers,  one for each of the populations in the model in
the same order as populations listed in the model. The values 
specify the down sampling to be used for each respective population.
For a population with k&gt;=1 diploid individuals (2k&gt;=2 genomes) in the model,
the downsample count d  must be 2&lt;=d&lt;=2k.</p>
</dd>
<dt><strong>--folded</strong> <em>&lt;True/False&gt;</em></dt><dd><p>The folded option indicates that the folded sfs should be returned.
If folded is False (default) the sfs reports the count of the derived allele.
If True,  the sfs reports of the count of the minor (less frequent) allele.</p>
</dd>
<dt><strong>--randomsnpprop</strong> <em>&lt;floating point value between 0 and 1&gt;</em></dt><dd><p>This option can be used to randomly sample a subset of SNPs. The default
is to sample all biallelic SNPs.</p>
</dd>
<dt><strong>--seed</strong> <em>&lt;integer&gt;</em></dt><dd><p>This is used with --randomsnpprop as the seed for the random number generator.</p>
</dd>
<dt><strong>--makeint</strong> <em>&lt;True/False&gt;</em></dt><dd><p>If True, round the counts in the sfs to the nearest integer (default False)</p>
</dd>
</dl>
</div>
<div class="section" id="example-usage">
<h2>Example usage<a class="headerlink" href="#example-usage" title="Permalink to this headline">¶</a></h2>
<p>Example command-lines:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>vcf_to_sfs.py -h
</pre></div>
</div>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>vcf_to_sfs.py --vcf pan_example2.vcf.gz --model-file panmodels.model --modelname 4Pop --downsamplesizes <span class="m">3</span> <span class="m">3</span> <span class="m">3</span> <span class="m">4</span> --folded --outgroup-fasta  chr22_pan_example2_ref.fa --out vcf_to_sfs_test1.txt
</pre></div>
</div>
</div>
<div class="section" id="importing-functions">
<h2>Importing Functions<a class="headerlink" href="#importing-functions" title="Permalink to this headline">¶</a></h2>
<blockquote>
<div><p>This file has two functions that can be useful for working with site frequency spectra</p>
</div></blockquote>
<div class="section" id="build-sfs">
<h3>build_sfs()<a class="headerlink" href="#build-sfs" title="Permalink to this headline">¶</a></h3>
<p>The default script that runs when this file is run. This function can also be
accessed directly by importing this file.</p>
<blockquote>
<div><ul class="simple">
<li><p>generates an sfs from a vcf file</p></li>
<li><p>can handle an arbitrary number of dimensions (populations) and
sample sizes, so long as each population has at least a
sample size of two genomes (i.e. one diploid individual)</p></li>
<li><p>handles downsampling, and reduction of dimensions</p></li>
<li><p>handles unfolded and folded sfs's</p></li>
<li><p>can take a BED file name to sample portions of a vcf file</p></li>
<li><p>can handle an alternative reference genome for rooting, rather
than that used in the vcf file</p></li>
<li><p>the arguments closely resemble those used when the function is called
by running this file</p></li>
</ul>
</div></blockquote>
<div class="section" id="id1">
<h4>Required Arguments<a class="headerlink" href="#id1" title="Permalink to this headline">¶</a></h4>
<blockquote>
<div><p>The first three arguments are, in order, the vcf filename,  the model file name, and the name of the model.
These are each strings.</p>
</div></blockquote>
</div>
<div class="section" id="optional-arguments">
<h4>Optional arguments<a class="headerlink" href="#optional-arguments" title="Permalink to this headline">¶</a></h4>
<blockquote>
<div><p>Each optional argument requires the use of the argument name.</p>
</div></blockquote>
<dl class="simple">
<dt><strong>BEDfilename</strong></dt><dd><p>The name of a ucsc-style bedfile with intervals to include</p>
</dd>
<dt><strong>altreference</strong></dt><dd><p>The name of a fasta sequence file that contains the reference genome</p>
</dd>
<dt><strong>folded</strong></dt><dd><p>True/False. To indicate that the folded sfs should be returned. (False is default)</p>
</dd>
<dt><strong>downsamplesizes</strong></dt><dd><p>A list of sample sizes to be used if they are
less than given in the model 2 &lt;= downsamplesizes[i] &lt;= samplesizes[i]</p>
</dd>
<dt><strong>randomsnpprop</strong></dt><dd><p>The proportion of snps to include using random sampling</p>
</dd>
<dt><strong>seed</strong></dt><dd><p>A random number seed that can be used with randomsnpprop.</p>
</dd>
<dt><strong>makeint</strong></dt><dd><p>Causes the array to be rounded to the nearest integer (dtype remains float)</p>
</dd>
<dt><strong>out</strong></dt><dd><p>The name of a file to contain the sfs
if out is not None,  this will write a tab-delimited file of the array</p>
</dd>
</dl>
</div>
<div class="section" id="id2">
<h4>Example usage<a class="headerlink" href="#id2" title="Permalink to this headline">¶</a></h4>
<p>Example python code:</p>
<div class="highlight-python notranslate"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre>1
2
3
4
5</pre></div></td><td class="code"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">vcf_to_sfs</span> <span class="kn">as</span> <span class="nn">vs</span>
<span class="n">mysfs</span> <span class="o">=</span> <span class="n">vs</span><span class="o">.</span><span class="n">build_sfs</span><span class="p">(</span><span class="n">pan_example2</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span><span class="p">,</span><span class="n">panmodels</span><span class="o">.</span><span class="n">model</span><span class="p">,</span><span class="s1">&#39;4Pop&#39;</span><span class="p">,</span>
        <span class="n">folded</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span><span class="n">downsamplesizes</span><span class="o">=</span><span class="p">[</span><span class="mi">3</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">],</span>
        <span class="n">altreference</span><span class="o">=</span><span class="s1">&#39;chr22_pan_example2_ref.fa&#39;</span><span class="p">,</span>
        <span class="n">out</span> <span class="o">=</span> <span class="s1">&#39;mysfsfile.txt&#39;</span><span class="p">)</span>
</pre></div>
</td></tr></table></div>
</div>
</div>
<div class="section" id="reduce-sfs-dims">
<h3>reduce_sfs_dims()<a class="headerlink" href="#reduce-sfs-dims" title="Permalink to this headline">¶</a></h3>
<blockquote>
<div><p>This function is for reducing the dimensionality of an sfs by summing across axes. It is
accessed by importing this file.</p>
<p>There are three required arguments in order:</p>
<ul class="simple">
<li><p>the sfs (i.e. a numpy array with as many dimensions as populations)</p></li>
<li><p>an instance of Class model that specifies the populations and samples to which the
sfs corresponds.</p></li>
<li><p>a list of names of the populations to keep in the reduced sfs</p></li>
</ul>
<p>There is one optional argument, 'out', if the reduced sfs is to be written to a file (e.g. out=mysfs.txt)</p>
</div></blockquote>
<div class="section" id="id3">
<h4>Example usage<a class="headerlink" href="#id3" title="Permalink to this headline">¶</a></h4>
<p>Example python code:</p>
<div class="highlight-python notranslate"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre>1
2
3</pre></div></td><td class="code"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">vcf_to_sfs</span> <span class="kn">as</span> <span class="nn">vs</span>
<span class="n">myreducedsfs</span> <span class="o">=</span> <span class="n">vs</span><span class="o">.</span><span class="n">reduce_sfs_dim</span><span class="p">(</span><span class="n">mysfs</span><span class="p">,</span><span class="n">mypopmodel</span><span class="p">,</span>
        <span class="p">[</span><span class="s1">&#39;A&#39;</span><span class="p">,</span><span class="s1">&#39;B&#39;</span><span class="p">,</span><span class="s1">&#39;C&#39;</span><span class="p">],</span><span class="n">out</span><span class="o">=</span><span class="s2">&quot;myreducedsfs_file.txt&quot;</span><span class="p">)</span>
</pre></div>
</td></tr></table></div>
</div>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="../utilities.html" class="btn btn-neutral float-right" title="PPP Utilities" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="plink_linkage_disequilibrium.html" class="btn btn-neutral float-left" title="plink_linkage_disequilibrium.py: Linkage Disequilibrium Analysis" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2020, PPP Development Team. This material is based upon work supported by the National Science Foundation under Grant Number (NSF ABI 1564659 to Arun Sethuraman and Jody Hey)

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>